torch/csrc/autograd/record_function.h

```diff
@@ -26,31 +28,79 @@ struct TORCH_API StringView {
   const char* str_ptr_;
 };
 
-using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
+enum class RecordFunctionState {
+  NO_CALLBACKS, // no callbacks registered, no need to use RecordFunction
+  NAME_ONLY, // callbacks need only basic information (function name)
+  NAME_AND_INPUTS, // callbacks need both name and function inputs
+};
+TORCH_API RecordFunctionState recordFunctionState();
 
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn, GetPackedInputsCallback cb = nullptr);
+  // Default constructor is used with before function called afterwards
+  RecordFunction() : state_(recordFunctionState()) {}
+
+  // Constructors with parameters initialize RecordFunction
+  // members and call start callbacks
+  explicit RecordFunction(Function* fn);
 
   explicit RecordFunction(
       std::string name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+      int64_t current_sequence_nr = -1);
 
   explicit RecordFunction(
       const char* name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+      int64_t current_sequence_nr = -1);
+
+  // before function initializes RecordFunction members and calls
+  // start callbacks;
+  // before function can be used only when RecordFunction is
+  // constructed with default constructor RecordFunction::RecordFunction()
+  void before(const char* name, int64_t sequence_nr = -1);
+  void before(std::string name, int64_t sequence_nr = -1);
+  void before(Function* fn, int64_t sequence_nr = -1);
+
+  template<typename F, typename... Args>
+  void before(
+      F fn,
+      int64_t current_sequence_nr,
+      const Args&... args) {
+    if (state_ == RecordFunctionState::NAME_AND_INPUTS) {
+      inputs_ = std::vector<c10::IValue>({args...});
+    }
+    before(fn, current_sequence_nr);
+  }
 
-  explicit RecordFunction(
-      std::string name,
-      GetPackedInputsCallback cb) : RecordFunction(name, -1, cb) {}
+  template<typename F, typename... Args>
+  void before(
+      F fn,
+      const c10::IValue& value,
```

Comment:
@zdevito, regarding your comment on "constructing IValues whenever any profiling is enabled".
I see this place is the one where we explicitly convert input to `c10::IValue`- when we pass a number of inputs (which are not `ArrayRef` or iterators) and do not specify sequence number, then we convert the first input to IValue. This conversion to IValue here is not necessary here yes, and can be improved further.

In other calls, correct me if I'm wrong, it should be fine - e.g. in the templated `before` function above compiler should generate `before` that explicitly accepts const references to whatever inputs we provided from the callsites and these inputs are then converted to `std::vector<c10::IValue>` but only if we are in `NAME_AND_INPUTS` state.